content/docs/quickstart/cpp.md

```diff
@@ -3,69 +3,163 @@ layout: quickstart
 
 ```sh
-$ cd third_party/protobuf
-$ make && sudo make install
+$ export PATH="$PATH:$MY_INSTALL_DIR/bin"
 ```
 
-###  Build the example
+### Prerequisites
 
-Always assuming you have gRPC properly installed, go into the example's
-directory:
+#### cmake
+
+Version 3.13 or later of `cmake` is required to install gRPC locally.
+
+- Linux
+
+  ```sh
+  $ sudo apt install -y cmake
+  ```
+
+- macOS:
+
+  ```sh
+  $ brew install cmake
+  ```
+
+- For general `cmake` installation instructions, see [Installing CMake][].
+
+Check the version of `cmake`:
 
 ```sh
-$ cd examples/cpp/helloworld/
+$ cmake --version
 ```
 
-Let's build the example client and server:
+Under Linux, the version of the system-wide `cmake` can be too low. You can
+install a more recent version into your local installation directory as follows:
+
 ```sh
-$ make
+$ wget -q -O cmake-linux.sh https://github.com/Kitware/CMake/releases/download/v{{< param cmake-version >}}/cmake-{{< param cmake-version >}}-Linux-x86_64.sh
+$ sh cmake-linux.sh -- --skip-license --prefix=$MY_INSTALL_DIR
+$ rm cmake-linux.sh
 ```
 
-Most failures at this point are a result of a faulty installation (or having
-installed gRPC to a non-standard location. Check out [the installation
-instructions for details](https://github.com/grpc/grpc/blob/master/src/cpp/README.md#make)).
+
+#### gRPC and Protocol Buffers v3
+
+While not mandatory, gRPC applications usually leverage [Protocol Buffers
+v3][pbv3] for service definitions and data serialization, and the example code
+uses Protocol Buffers.
+
+The following instructions will locally install gRPC and Protocol Buffers.
+
+ 1. Install the basic tools required to build gRPC:
+
+    - Linux
+
+      ```sh
+      $ sudo apt install -y build-essential autoconf libtool pkg-config
+      ```
+
+    - macOS:
+
+      ```sh
+      $ brew install autoconf automake libtool pkg-config
+      ```
+
+ 2. Clone the `grpc` repo and its submodules:
+
+    ```sh
+    $ git clone --recurse-submodules -b {{< param grpc_release_tag >}} https://github.com/grpc/grpc
+    $ cd grpc
+    ```
+
+ 3. Build and locally install gRPC and all requisite tools:
+
+    ```sh
+    $ mkdir -p cmake/build
+    $ pushd cmake/build
+    $ cmake -DgRPC_INSTALL=ON \
+          -DgRPC_BUILD_TESTS=OFF \
+          -DCMAKE_INSTALL_PREFIX=$MY_INSTALL_DIR \
+          ../..
+    $ make -j
+    $ make install
+    $ popd
+    ```
+
+You can find a complete set of instructions for building gRPC C++ in [Building
+from source][from-source].
+
+[from-source]: https://github.com/grpc/grpc/blob/master/BUILDING.md
+
+###  Build the example
+
+The example code is part of the `grpc` repo source, which you cloned as part of
+the steps of the previous section.
+
+ 1. Change to the example's directory:
+
+    ```sh
+    $ cd examples/cpp/helloworld
+    ```
+
+ 2. Build the example using `cmake`:
+
+    ```sh
+    $ mkdir -p cmake/build
+    $ pushd cmake/build
+    $ cmake ../..
```

Comment:
My understanding is that `CMAKE_INSTALL_PREFIX` is to be used when we want to install something from a package, but we aren't installing anything from the example.

I've retested the setup instructions and they work as given, provided the `PATH` variable is properly set. Recall that one of the steps earlier in the page is:

```console
$ export PATH="$PATH:$MY_INSTALL_DIR/bin"
```

If I omit this step, then running `cmake` on the example fails:

```console
cmake ../..
-- The C compiler identification is GNU 7.5.0
...
CMake Error at CMakeLists.txt:103 (find_package):
  Could not find a package configuration file provided by "Protobuf" with any
  of the following names:

    ProtobufConfig.cmake
    protobuf-config.cmake

  Add the installation prefix of "Protobuf" to CMAKE_PREFIX_PATH or set
  "Protobuf_DIR" to a directory containing one of the above files.  If
  "Protobuf" provides a separate development package or SDK, be sure it has
  been installed.
-- Configuring incomplete, errors occurred!
...
```

After I set the `PATH` variable, then `cmake ../..` finds `protoc` and `grpc`:

```console
cmake ../..
CMake Warning (dev) at /home/chalin/local/lib/cmake/protobuf/protobuf-options.cmake:6 (option):
  Policy CMP0077 is not set: option() honors normal variables.  ...
-- Using protobuf
-- Using gRPC 1.28.1
-- Configuring done
-- Generating done
-- Build files have been written to: /home/chalin/grpc/examples/cpp/helloworld/cmake/build
```

Reply:
I feel that setting PATH variable is dirty and it's not the right cmake way of doing it.

Actually one of the error messages you posted gives a hint about what to do (*Add the installation prefix of "Protobuf" to CMAKE_PREFIX_PATH or set
  "Protobuf_DIR" to a directory containing one of the above files*).

In one of our distribtests we're actually just setting CMAKE_INSTALL_PREFIX and it works just fine
https://github.com/grpc/grpc/blob/80b74811d466db2f7cfed06f0f3b33395b15c876/test/distrib/cpp/run_distrib_test_cmake.bat#L86

If everything works after setting CMAKE_PREFIX_PATH, I'd use that because it seems to be the cleanest option.
https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html

Reply:
Yes it works with `CMAKE_PREFIX_PATH`. I've updated the command to make use of it.